File: FateGrandAutomata.Core/Modules/Card.cs
using System.Collections.Generic;
using CoreAutomata;

namespace FateGrandAutomata
{
    public class Card
    {
        public AutoSkill AutoSkill { get; private set; }
        public Battle Battle { get; private set; }

        readonly List<CardScore> _cardPriorityArray = new List<CardScore>();

        Dictionary<CardScore, List<int>> _commandCards = new Dictionary<CardScore, List<int>>();
        int _cardsClickedSoFar;

        bool chains;

        public void Init(AutoSkill AutoSkillModule, Battle BattleModule)
        {
            AutoSkill = AutoSkillModule;
            Battle = BattleModule;

            InitCardPriorityArray();
        }

        void InitCardPriorityArray()
        {
            var errorString = "Battle_CardPriority Error at '";

            if (Preferences.Instance.BattleCardPriority.Length == 3)
            {
                InitCardPriorityArraySimple(errorString);
            }
            else InitCardPriorityArrayDetailed(errorString);
        }

        void InitCardPriorityArraySimple(string ErrorString)
        {
            foreach (var card in Preferences.Instance.BattleCardPriority)
            {
                var score = CardScore.Normal;

                switch (card)
                {
                    case 'B':
                        score = CardScore.Buster;
                        break;

                    case 'A':
                        score = CardScore.Arts;
                        break;

                    case 'Q':
                        score = CardScore.Quick;
                        break;

                    default:
                        throw new ScriptExitException($"{ErrorString}{card}': Only 'B', 'A' and 'Q' are allowed in simple mode.");
                }

                _cardPriorityArray.Add(score | CardScore.Weak);
                _cardPriorityArray.Add(score);
                _cardPriorityArray.Add(score | CardScore.Resist);
            }
        }

        void InitCardPriorityArrayDetailed(string ErrorString)
        {
            var cardCounter = 0;

            foreach (var i in Preferences.Instance.BattleCardPriority.Split(','))
            {
                var card = i.ToUpper().Trim();

                if (card.Length < 1 || card.Length > 2)
                {
                    throw new ScriptExitException($"{ErrorString}{card}': Invalid card length.");
                }

                var score = CardScore.Normal;

                switch (card.Length == 1 ? card[0] : card[1])
                {
                    case 'B':
                        score |= CardScore.Buster;
                        break;

                    case 'A':
                        score |= CardScore.Arts;
                        break;

                    case 'Q':
                        score |= CardScore.Quick;
                        break;

                    default:
                        throw new ScriptExitException($"{ErrorString}{card}': Only 'B', 'A' and 'Q' are valid card types.");
                }

                if (card.Length == 2)
                {
                    switch (card[0])
                    {
                        case 'W':
                            score |= CardScore.Weak;
                            break;

                        case 'R':
                            score |= CardScore.Resist;
                            break;

                        default:
                            throw new ScriptExitException($"{ErrorString}{card}': Only 'W', and 'R' are valid card affinities.");
                    }
                }

                _cardPriorityArray.Add(score);
                ++cardCounter;
            }

            if (cardCounter != 9)
            {
                throw new ScriptExitException($"{ErrorString}{Preferences.Instance.BattleCardPriority}': Expected 9 cards, but {cardCounter} found.");
            }
        }

        CardScore GetCardAffinity(Region Region)
        {
            if (Region.Exists(ImageLocator.Weak))
            {
                return CardScore.Weak;
            }

            if (Region.Exists(ImageLocator.Resist))
            {
                return CardScore.Resist;
            }

            return CardScore.Normal;
        }

        CardScore GetCardType(Region Region)
        {
            if (Region.Exists(ImageLocator.Buster))
            {
                return CardScore.Buster;
            }

            if (Region.Exists(ImageLocator.Art))
            {
                return CardScore.Arts;
            }

            if (Region.Exists(ImageLocator.Quick))
            {
                return CardScore.Quick;
            }

            AutomataApi.Toast($"Failed to determine Card type (X: {Region.X}, Y: {Region.Y}, W: {Region.W}, H: {Region.H})");

            return CardScore.Buster;
        }

        Dictionary<CardScore, List<int>> GetCommandCards()
        {
            var storagePerPriority = new Dictionary<CardScore, List<int>>();

            AutomataApi.UseSameSnapIn(() =>
            {
                for (var cardSlot = 0; cardSlot < 5; ++cardSlot)
                {
                    var score = GetCardAffinity(Game.BattleCardAffinityRegionArray[cardSlot])
                                | GetCardType(Game.BattleCardTypeRegionArray[cardSlot]);

                    if (!storagePerPriority.ContainsKey(score))
                    {
                        storagePerPriority.Add(score, new List<int>());
                    }

                    storagePerPriority[score].Add(cardSlot);
                }
            });

            return storagePerPriority;
        }

        public void ClickCommandCards(int Clicks)
        {
            if (chains)
            {
                return;
            }

            var i = 1;

            foreach (var cardPriority in _cardPriorityArray)
            {
                if (!_commandCards.ContainsKey(cardPriority))
                    continue;

                var currentCardTypeStorage = _commandCards[cardPriority];

                foreach (var cardSlot in currentCardTypeStorage)
                {
                    if (Clicks < i)
                    {
                        _cardsClickedSoFar = i - 1;
                        return;
                    }

                    if (i > _cardsClickedSoFar)
                    {
                        Game.BattleCommandCardClickArray[cardSlot].Click();
                    }

                    ++i;
                }
            }
        }

        public bool CanClickNpCards
        {
            get
            {
                var weCanSpam = Preferences.Instance.BattleNoblePhantasm == BattleNoblePhantasmType.Spam;
                var weAreInDanger = Preferences.Instance.BattleNoblePhantasm == BattleNoblePhantasmType.Danger
                                    && Battle.HasChoosenTarget;

                return (weCanSpam || weAreInDanger) && AutoSkill.IsFinished;
            }
        }

        public bool ClickNpCards()
        {
            var npsClicked = false;

            foreach (var npCard in Game.BattleNpCardClickArray)
            {
                npCard.Click();

                npsClicked = true;
            }

            return npsClicked;
        }

        public void ReadCommandCards()
        {
            _commandCards = GetCommandCards();
        }

        public void ResetCommandCards()
        {
            _commandCards = new Dictionary<CardScore, List<int>>();
            _cardsClickedSoFar = 0;
        }
    }
}

File: app/src/main/java/com/mathewsachin/fategrandautomata/scripts/modules/Card.kt
package com.mathewsachin.fategrandautomata.scripts.modules

import com.mathewsachin.fategrandautomata.core.*
import com.mathewsachin.fategrandautomata.scripts.CardScore
import com.mathewsachin.fategrandautomata.scripts.ImageLocator
import com.mathewsachin.fategrandautomata.scripts.enums.BattleNoblePhantasmEnum
import com.mathewsachin.fategrandautomata.scripts.enums.CardAffinityEnum
import com.mathewsachin.fategrandautomata.scripts.enums.CardTypeEnum
import com.mathewsachin.fategrandautomata.scripts.prefs.Preferences

private const val dummyNormalAffinityChar = 'X'
private const val cardPriorityErrorString = "Battle_CardPriority Error at '"

fun getCardScores(Priority: String): List<CardScore> {
    val scores = Priority
        .splitToSequence(',')
        .map { it.trim().toUpperCase() }
        .map {
            when (it.length) {
                1 -> "${dummyNormalAffinityChar}$it"
                2 -> it
                else -> throw ScriptExitException("${cardPriorityErrorString}${it}': Invalid card length.")
            }
        }
        .map {
            val cardType = when(it[1]) {
                'B' -> CardTypeEnum.Buster
                'A' -> CardTypeEnum.Arts
                'Q' -> CardTypeEnum.Quick
                else -> throw ScriptExitException("${cardPriorityErrorString}${it[1]}': Only 'B', 'A' and 'Q' are valid card types.")
            }

            val cardAffinity = when(it[0]) {
                'W' -> CardAffinityEnum.Weak
                'R' -> CardAffinityEnum.Resist
                dummyNormalAffinityChar -> CardAffinityEnum.Normal
                else -> throw ScriptExitException("${cardPriorityErrorString}${it[0]}': Only 'W', and 'R' are valid card affinities.")
            }

            CardScore(cardType, cardAffinity)
        }
        .toList()

    if (scores.size != 9) {
        throw ScriptExitException("${cardPriorityErrorString}': Expected 9 cards, but ${scores.size} found.")
    }

    return scores
}

class Card {
    private lateinit var autoSkill: AutoSkill
    private lateinit var battle: Battle

    private lateinit var cardPriorityArray: List<CardScore>

    private val commandCards = mutableMapOf<CardScore, MutableList<Int>>()
    private var cardsClickedSoFar = 0

    fun init(AutoSkillModule: AutoSkill, BattleModule: Battle) {
        autoSkill = AutoSkillModule
        battle = BattleModule

        initCardPriorityArray()
    }

    private fun initCardPriorityArray() {
        val priority = Preferences.BattleCardPriority
        if (priority.length == 3) {
            initCardPriorityArraySimple(priority)
        }
        else initCardPriorityArrayDetailed(priority)
    }

    private fun initCardPriorityArraySimple(Priority: String) {
        val detailedPriority = Priority
            .map { "W$it, ${dummyNormalAffinityChar}$it, R$it" }
            .joinToString()

        initCardPriorityArrayDetailed(detailedPriority)
    }

    private fun initCardPriorityArrayDetailed(Priority: String) {
        cardPriorityArray = getCardScores(Priority)
    }

    private fun getCardAffinity(Region: Region): CardAffinityEnum {
        if (Region.exists(ImageLocator.Weak)) {
            return CardAffinityEnum.Weak
        }

        if (Region.exists(ImageLocator.Resist)) {
            return CardAffinityEnum.Resist
        }

        return CardAffinityEnum.Normal
    }

    private fun getCardType(Region: Region): CardTypeEnum {
        if (Region.exists(ImageLocator.Buster)) {
            return CardTypeEnum.Buster
        }

        if (Region.exists(ImageLocator.Art)) {
            return CardTypeEnum.Arts
        }

        if (Region.exists(ImageLocator.Quick)) {
            return CardTypeEnum.Quick
        }

        AutomataApi.toast("Failed to determine Card type $Region")

        return CardTypeEnum.Buster
    }

    fun readCommandCards() {
        commandCards.clear()

        AutomataApi.useSameSnapIn {
            for (cardSlot in 0..4) {
                val affinity = getCardAffinity(Game.BattleCardAffinityRegionArray[cardSlot])
                val type = getCardType(Game.BattleCardTypeRegionArray[cardSlot])

                val score = CardScore(type, affinity)

                if (!commandCards.containsKey(score)) {
                    commandCards[score] = mutableListOf()
                }

                commandCards[score]?.add(cardSlot)
            }
        }
    }

    val canClickNpCards: Boolean get() {
        val weCanSpam = Preferences.BattleNoblePhantasm == BattleNoblePhantasmEnum.Spam
        val weAreInDanger = Preferences.BattleNoblePhantasm == BattleNoblePhantasmEnum.Danger
                && battle.hasChosenTarget

        return (weCanSpam || weAreInDanger) && autoSkill.isFinished
    }

    fun clickNpCards() {
        for (npCard in Game.BattleNpCardClickArray) {
            npCard.click()
        }
    }

    fun clickCommandCards(Clicks: Int) {
        var i = 1

        for (cardPriority in cardPriorityArray) {
            if (!commandCards.containsKey(cardPriority))
                continue

            val currentCardTypeStorage = commandCards[cardPriority]
                ?: continue

            for (cardSlot in currentCardTypeStorage) {
                if (Clicks < i) {
                    cardsClickedSoFar = i - 1
                    return
                }

                if (i > cardsClickedSoFar) {
                    Game.BattleCommandCardClickArray[cardSlot].click()
                }

                ++i
            }
        }
    }

    fun resetCommandCards() {
        commandCards.clear()
        cardsClickedSoFar = 0
    }
}